src/snowflake/snowpark/_internal/ast/batch.py

```diff
@@ -79,14 +107,15 @@ def bind(self, symbol: Optional[str] = None) -> proto.Bind:
         Args:
             symbol: An optional symbol to name the new variable.
         """
-        stmt = self._request.body.add()
-        # TODO: extended BindingId spec from the branch snowpark-ir.
-        stmt.bind.first_request_id = (
-            self._request_id.bytes
-        )  # Set the first batch ID this stmt will be flushed with.
-        stmt.bind.uid = self._get_next_id()
-        stmt.bind.symbol.value = symbol if isinstance(symbol, str) else ""
-        return stmt.bind
+        with self._lock:
```

Comment:
put the lock only on 
```
stmt.bind.first_request_id = (
                self._request_id.bytes
            )  # Set the first batch ID this stmt will be flushed with.
stmt.bind.uid = next(AstBatch.__id_gen)
self._cur_stmts.append(stmt)
```
This is what only needs to be guarded.

A more elegant solution may be to simply use a helper function with a lock, e.g.
```
return self._append_stmt(stmt)
```
and then have
```
def _append_stmt(self, stmt: proto.Stmt) -> proto.Bind:
    with self._lock:
         stmt.bind.first_request_id = self._request_id.bytes
         stmt.bind.uid = next(AstBatch.__id_gen)
         self._cut_stmt.append(stmt)
         return stmt.bind
```

Reply:
I think we should just maintain our own `RLock` instance. I'm weary of issues even if thread safe sessions aren't enabled due to the possibility that multiple users acquire the same `Session` instance through whatever means, and additionally reusing the `Session` instance's own lock seems like we're inviting future trouble as more and more attributes make use of it.

Reply:
ok, sgtm.